### PR TITLE
Refactor Dune_file.Stanzas.parser and clean up Dune_init stanza conflict detection

### PR DESCRIPTION
**Pull request by shonfeder**
Fulfills @rgrinberg's request at https://github.com/ocaml/dune/pull/1448#issuecomment-472710046.

The new auxiliary function and light refactor in e24ef68 follows lines discussed with @diml in  #1972

### CODE REVIEW

File: src/dune_init.ml [hunk 1]
```diff
@@ -55,76 +55,56 @@ module File = struct
...
-        || strict_equal a.public_name b.public_name
-    end
-
     let pp ppf s =
       Option.iter (Cst.to_sexp s) ~f:(Dune_lang.pp Dune ppf)
 
+    let libraries_conflict (a: Dune_file.Library.t) (b: Dune_file.Library.t) =
+      a.name = b.name
+
+    module StringSet = Set.Make(String)
```
Comment (rgrinberg) [inline src/dune_init.ml]: We have `String.Set` already

Comment (shonfeder) [reply src/dune_init.ml]: Thanks for the tip!

**Comment by shonfeder** [comment]
Addressed comment and rebased.

**Comment by shonfeder** [comment]
Rebased again to bring the code up to date.

**Comment by rgrinberg** [comment]
I merged with a couple of minor tweaks. Thanks.

**Pull request merged**